javascript/ql/src/semmle/javascript/security/dataflow/XssThroughDomCustomizations.qll

```diff
@@ -51,14 +58,32 @@ module XssThroughDom {
     }
   }
 
+  /**
+   * A source for text from the DOM from a `d3` method call.
+   */
+  class D3TextSource extends Source {
+    D3TextSource() {
+      exists(DataFlow::MethodCallNode call, string methodName, string argValue |
+        this = call and
+        call = D3::d3Selection().getMember(methodName).getACall() and
+        call.getNumArgument() = 1 and
+        call.getArgument(0).mayHaveStringValue(argValue)
+      |
+        methodName = "attr" and argValue = unsafeAttributeName()
+        or
+        methodName = "property" and argValue = unsafeDomPropertyName()
+      )
+    }
+  }
+
   /**
    * A source for text from the DOM from a DOM property read or call to `getAttribute()`.
    */
   class DOMTextSource extends Source {
     DOMTextSource() {
       exists(DataFlow::PropRead read | read = this |
         read.getBase().getALocalSource() = DOM::domValueRef() and
-        read.mayHavePropertyName(["innerText", "textContent", "value", "name"])
+        read.mayHavePropertyName(unsafeDomPropertyName())
```

Comment:
Unrelated to D3, but something I just thought of: 
The [`$().prop(propName)`](https://api.jquery.com/prop/) method might also be worth including here.

Reply:
Adding it in here and running another evaluation.